src/store.rs

```diff
@@ -0,0 +1,196 @@
+                format!("could not read `{}` to calculate hash", source.display())
+            })?;
+
+            ///////////////////////////////////////////////
+            // Step 3: make sure any parent paths exist  //
+            ///////////////////////////////////////////////
+            let mut ancestors: Vec<&Path> = source.ancestors().skip(1).collect();
+            ancestors.pop(); // we've made sure this is relative, so the first item is ""
+            ancestors.reverse(); // go `[a, a/b, a/b/c]` instead of `[a/b/c, a/b, a]`
+
+            for ancestor_path in ancestors {
+                let ancestor = ancestor_path.to_path_buf();
+
+                if output_dirs.contains(&ancestor) {
+                    continue;
+                }
+
+                log::trace!(
+                    "creating {} in {}",
+                    &ancestor.display(),
+                    temp.path().display()
+                );
+                std::fs::create_dir(temp.path().join(&ancestor)).with_context(|| {
+                    format!(
+                        "could not create ancestor `{}` for output `{}`",
+                        ancestor.display(),
+                        source.display(),
+                    )
+                })?;
+                output_dirs.insert(ancestor);
+            }
+
+            //////////////////////////////
+            // Step 4: collect the file //
+            //////////////////////////////
+            log::trace!("moving {} into collection path", &source.display());
+            let out = temp.path().join(&source);
+            std::fs::rename(workspace.join(&source), &out).with_context(|| {
+                format!(
+                    "could not move `{}` from workspace to store",
+                    source.display()
+                )
+            })?;
+
+            Self::make_readonly(&out).with_context(|| {
+                format!(
+                    "could not make `{}` read-only after moving into store",
+                    out.display()
+                )
+            })?;
+        }
+
+        // Finish up by making sure everything is readonly and store it in the
+        // final location.
+        for dir in &output_dirs {
+            Self::make_readonly(&temp.path().join(&dir)).with_context(|| {
+                format!("could not make `{}` read-only in the store", dir.display(),)
+            })?;
+        }
+
+        let final_hash = hasher.finalize().to_hex().to_string();
+        let final_location = self.root.join(&final_hash);
+
+        log::debug!("moving {} into store", &final_hash);
+
+        // important: at this point we need to take ownership of the tempdir so
+        // that it doesn't get automatically removed when it's dropped. We've
+        // so far avoided that to avoid leaving temporary directories laying
+        // around in case of errors.
+        std::fs::rename(temp.into_path(), &final_location)
```

Comment:
Instead of having a temporary directory that is in `/tmp`, would it make more sense to put in the store with a prefix or suffix, that way if something goes wrong, you can still look at what was outputted and potentially debug better?

Reply:
Yeah, that might make sense. Easy to implement, too!